src/TurbulenceClosures/turbulence_closure_implementations/tke_based_vertical_diffusivity.jl

```diff
@@ -0,0 +1,370 @@
+
+        return new{TD, CK, CD, CL, CQ}(diffusivity_scaling, dissipation_parameter, mixing_length_parameter, surface_model)
+    end
+end
+
+"""
+    TKEBasedVerticalDiffusivity <: AbstractTurbulenceClosure{TD}
+
+Parameters for the "anisotropic minimum dissipation" turbulence closure for large eddy simulation
+proposed originally by [Rozema15](@cite) and [Abkar16](@cite), and then modified
+by [Verstappen18](@cite), and finally described and validated for by [Vreugdenhil18](@cite).
+"""
+function TKEBasedVerticalDiffusivity(FT=Float64;
+                                     diffusivity_scaling = RiDependentDiffusivityScaling{FT}(),
+                                     dissipation_parameter = 2.91,
+                                     mixing_length_parameter = 1.16,
+                                     surface_model = TKESurfaceFlux{FT}(),
+                                     time_discretization::TD = ExplicitTimeDiscretization()) where TD
+
+    dissipation_parameter = convert(FT, dissipation_parameter)
+    mixing_length_parameter = convert(FT, mixing_length_parameter)
+    diffusivity_scaling = convert_eltype(FT, diffusivity_scaling)
+    surface_model = convert_eltype(FT, surface_model)
+
+    return TKEBasedVerticalDiffusivity{TD}(diffusivity_scaling,
+                                           dissipation_parameter,
+                                           mixing_length_parameter,
+                                           surface_model)
+end
+
+const TKEVD = TKEBasedVerticalDiffusivity
+
+"""
+    struct RiDependentDiffusivityScaling{FT}
+
+A diffusivity model in which momentum, tracers, and TKE
+each have Richardson-number-dependent diffusivities.
+The Richardson number is
+
+    ``Ri = ∂z B / ( (∂z U)² + (∂z V)² )`` ,
+
+where ``B`` is buoyancy and ``∂z`` denotes a vertical derviative.
+The Richardson-number dependent diffusivities are multiplied by the stability
+function
+
+    ``σ(Ri) = σ⁻ + (σ⁺ - σ⁻) * step(Ri, Riᶜ, Riʷ)``
+
+where ``σ⁰``, ``σᵟ``, ``Riᶜ``, and ``Riʷ`` are free parameters,
+and ``step`` is a smooth step function defined by
+
+    ``step(x, c, w) = (1 + tanh((x - c) / w)) / 2``.
+"""
+Base.@kwdef struct RiDependentDiffusivityScaling{FT}
+    Cᴷu⁻  :: FT = 0.15
+    Cᴷu⁺  :: FT = 0.73
+    Cᴷc⁻  :: FT = 0.40
+    Cᴷc⁺  :: FT = 1.77
+    Cᴷe⁻  :: FT = 0.13
+    Cᴷe⁺  :: FT = 1.22
+    CᴷRiʷ :: FT = 0.72
+    CᴷRiᶜ :: FT = 0.76
+end
+
+Base.@kwdef struct TKESurfaceFlux{FT}
+    Cᵂu★ :: FT = 3.62
+    CᵂwΔ :: FT = 1.31
+end
+
+for S in (:RiDependentDiffusivityScaling, :TKESurfaceFlux)
+    @eval @inline convert_eltype(::Type{FT}, s::$S) where FT = $S{FT}(; Dict(p => getproperty(s, p) for p in propertynames(s))...)
+    @eval @inline convert_eltype(::Type{FT}, s::$S{FT}) where FT = s
+end
+
+#####
+##### Utilities
+#####
+
+#
+# TODO: figure out how to calculate the current buoyancy flux and momentum flux...
+# function validate_closure_dependent_boundary_conditions(closure::TKEVD, boundary_conditions)
+#
+# end
```

Comment:
Still a TODO?

Reply:
Hmm yes. Getting this right requires a bunch of new code. Basically we need a `surface_buoyancy_flux(i, j, grid, buoyancy_model, tracer_boundary_conditions, other_args...)` utility... I guess I was a bit hasty in requesting review because this should probably be part of this PR, otherwise it's fairly hard to actually use `TKEBasedVerticalDiffusivity` correctly (you'd have to manually implement the correct surface flux of TKE, as I do in the validation experiments).